src/main/java/com/scalar/db/storage/jdbc/JDBCService.java

```diff
@@ -0,0 +1,213 @@
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+
+/**
+ * A service class to perform get/scan/put/delete/mutate operations
+ *
+ * @author Toshihiro Suzuki
+ */
+@ThreadSafe
+public class JDBCService {
+
+  private final OperationChecker operationChecker;
+  private final QueryBuilder queryBuilder;
+  private final String schemaPrefix;
+
+  public JDBCService(
+      OperationChecker operationChecker, QueryBuilder queryBuilder, String schemaPrefix) {
+    this.operationChecker = Objects.requireNonNull(operationChecker);
+    this.queryBuilder = Objects.requireNonNull(queryBuilder);
+    this.schemaPrefix = Objects.requireNonNull(schemaPrefix);
+  }
+
+  private Table getTable(
+      Operation operation, @Nullable String defaultSchema, @Nullable String defaultTable) {
+    String schema = operation.forNamespace().orElse(defaultSchema);
+    String table = operation.forTable().orElse(defaultTable);
+    return new Table(schemaPrefix, schema, table);
+  }
+
+  public Optional<Result> get(
+      Get get, Connection connection, @Nullable String defaultSchema, @Nullable String defaultTable)
+      throws SQLException {
+    addProjectionsForKeys(get);
+    Table table = getTable(get, defaultSchema, defaultTable);
+    @Nullable Key clusteringKey = get.getClusteringKey().orElse(null);
+
+    operationChecker.checkGet(table, get.getProjections(), get.getPartitionKey(), clusteringKey);
+
+    SelectQuery selectQuery =
+        queryBuilder
+            .select(get.getProjections())
+            .from(table)
+            .where(get.getPartitionKey(), clusteringKey)
+            .build();
+
+    PreparedStatement preparedStatement = selectQuery.prepareAndBind(connection);
+    try (ResultSet resultSet = preparedStatement.executeQuery()) {
+      if (resultSet.next()) {
+        return Optional.of(selectQuery.getResult(resultSet));
+      }
+      return Optional.empty();
+    }
+  }
+
+  public Scanner scan(
+      Scan scan,
+      Connection connection,
+      @Nullable String defaultSchema,
+      @Nullable String defaultTable)
+      throws SQLException {
+    addProjectionsForKeys(scan);
+    Table table = getTable(scan, defaultSchema, defaultTable);
+
+    @Nullable Key startClusteringKey = scan.getStartClusteringKey().orElse(null);
+    @Nullable Key endClusteringKey = scan.getEndClusteringKey().orElse(null);
+
+    operationChecker.checkScan(
+        table,
+        scan.getProjections(),
+        scan.getPartitionKey(),
+        startClusteringKey,
+        endClusteringKey,
+        scan.getLimit(),
+        scan.getOrderings());
+
+    SelectQuery selectQuery =
+        queryBuilder
+            .select(scan.getProjections())
+            .from(table)
+            .where(
+                scan.getPartitionKey(),
+                startClusteringKey,
```

Comment:
I feel, passing optional as it is seems much simpler than getting nullable key and doing null check in the method.
https://github.com/scalar-labs/scalardb/pull/133/files#diff-8810b9d8331d1df08af5790eacab666ecde1ee1523a8f6f1db3e893987dc0ebfR73

Is there any particular reason of this?

Reply:
As discussed, I will update the code to use Optional as a parameter and a class field.